File: 프로그래머스/lv3/43162. 네트워크/네트워크.py
def solution(n, computers):
    p = [i for i in range(n)]

    def find(p, x):
        if p[x] != x:
            p[x] = find(p,p[x])
        return p[x]

    def union(x, y):
        px = find(p,x)
        py = find(p,y)

        if px < py:
            p[py] = px
        else:
            p[px] = py

    for i in range(n):
        for j in range(n):
            if computers[i][j] == 1:
                union(i, j)

    for i in range(n):
        find(p, i)

    return len(set(p))

File: 프로그래머스/lv3/43162. 네트워크/네트워크.java
class Solution {
    public static boolean[] check;
    public int solution(int n, int[][] computers) {
        int answer = 0;
        check = new boolean[n];

        for (int i = 0; i < n; i++) {
            if (!check[i]) {
                answer++;
                dfs(i, n, computers);
            }
        }
        return answer;
    }

    public void dfs(int idx, int n, int[][] computers) {
        check[idx] = true;

        for (int i = 0; i < n; i++) {
            if (idx != i && computers[idx][i] == 1 && !check[i]) {
                dfs(i, n, computers);
            }
        }
    }
}